<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2018-01-03 Wed 16:44 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>Repeat Masking Pipeline</title>
<meta name="generator" content="Org mode" />
<meta name="author" content="Robert Syme" />
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="./theme.css" rel="stylesheet">
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">Repeat Masking Pipeline</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org2722ddc">1. Installing</a></li>
<li><a href="#orgb236b89">2. Walkthrough</a>
<ul>
<li><a href="#orgf3006e5">2.1. Introduction</a></li>
<li><a href="#org18191d0">2.2. Parameter Setup</a></li>
<li><a href="#orgafd5440">2.3. Collection of relatively recent LTR retrotranposons</a></li>
<li><a href="#org57db473">2.4. Collection of relatively old LTR retrotransposons</a></li>
<li><a href="#orgbe302cb">2.5. Cleaning LTR results</a></li>
<li><a href="#orgcae4ec4">2.6. Identify elements with nested insertions</a></li>
<li><a href="#orgeadaf53">2.7. Noncoding RNA</a></li>
<li><a href="#org41d7682">2.8. Final Masking</a></li>
<li><a href="#org478605c">2.9. Summary tables and figures</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org2722ddc" class="outline-2">
<h2 id="org2722ddc"><span class="section-number-2">1</span> Installing</h2>
<div class="outline-text-2" id="text-1">
<p>
Ideally, the only prerequisites would be <a href="https://www.nextflow.io/">Nextflow</a> and
<a href="https://www.docker.com/">Docker</a>. Unfortunately, it's almost impossible to do work on repeats
without the RepBase library which carries licence restrictions so
I'm unable to include the library in the Docker image.
</p>

<p>
To build your own image that <i>does</i> include the RepBase libraries,
run:
</p>

<div class="org-src-container">
<pre class="src src-sh"><span style="color: #c678dd;">cd</span> <span style="color: #51afef; font-weight: bold;">`mktemp -d`</span>

<span style="color: #5B6268;"># </span><span style="color: #5B6268;">Download the RepBase repeat library (replace RB_USERNAME and RB_PASSWORD with your username and password)</span>
wget --user $<span style="color: #dcaeea;">RB_USERNAME</span> <span style="color: #98be65;">\</span>
     --password $<span style="color: #dcaeea;">RB_PASSWORD</span> <span style="color: #98be65;">\</span>
     --output-document repeatmaskerlibraries.tar.gz <span style="color: #98be65;">\</span>
     http://www.girinst.org/server/RepBase/protected/repeatmaskerlibraries/repeatmaskerlibraries-20140131.tar.gz

<span style="color: #5B6268;"># </span><span style="color: #5B6268;">Make an (almost) empty Dockerfile.</span>
<span style="color: #5B6268;">#  </span><span style="color: #5B6268;">All of the important instructions are in the repeatmasker-onbuild image</span>
<span style="color: #5B6268;">#  </span><span style="color: #5B6268;">You can view them here: https://github.com/robsyme/nextflow-annotate/blob/master/Dockerfiles/RepeatMasker-onbuild/Dockerfile</span>
<span style="color: #c678dd;">echo</span> <span style="color: #98be65;">"FROM robsyme/nf-repeatmasking"</span> &gt; Dockerfile

<span style="color: #5B6268;"># </span><span style="color: #5B6268;">Build a new docker image called 'nf-repeatmasking'.</span>
<span style="color: #5B6268;">#  </span><span style="color: #5B6268;">When building, this image looks for a file called 'repeatmaskerlibraries.tar.gz' which it pulls into the image.</span>
docker build -t repeats .
</pre>
</div>
</div>
</div>

<div id="outline-container-orgb236b89" class="outline-2">
<h2 id="orgb236b89"><span class="section-number-2">2</span> Walkthrough</h2>
<div class="outline-text-2" id="text-2">
</div>
<div id="outline-container-orgf3006e5" class="outline-3">
<h3 id="orgf3006e5"><span class="section-number-3">2.1</span> Introduction</h3>
<div class="outline-text-3" id="text-2-1">
<p>
This is a <a href="https://en.wikipedia.org/wiki/Literate_programming">literate-programming</a> style walk through the pipeline,
explaining the utility of every section. Each code block below gets
tangled to compose the  pipline, which is written out to <a href="../main.nf"><code>main.nf</code></a>.
</p>
</div>
</div>
<div id="outline-container-org18191d0" class="outline-3">
<h3 id="org18191d0"><span class="section-number-3">2.2</span> Parameter Setup</h3>
<div class="outline-text-3" id="text-2-2">
<p>
There is some necessary setup and housekeeping to begin. This is
where we set up our strain name, the path to our reference sequence
and the output location for our final figures and tables.
</p>

<div class="org-src-container">
<pre class="src src-groovy">VERSION = <span style="color: #98be65;">"1.0.0"</span>

params.reference = <span style="color: #98be65;">'data/example/genome.fasta'</span>
params.trnaprot = <span style="color: #98be65;">'http://www.hrt.msu.edu/uploads/535/78637/Tpases020812.gz'</span>
params.trnanuc = <span style="color: #98be65;">'http://gtrnadb2009.ucsc.edu/download/tRNAs/eukaryotic-tRNAs.fa.gz'</span>
params.outdir = <span style="color: #98be65;">'output'</span>
params.minRepLength = 250

trnanuc = file(params.trnanuc)
trnaprot = file(params.trnaprot)
reference = file(params.reference)

log.info <span style="color: #98be65;">""</span>
log.info <span style="color: #98be65;">"R E P E A T M A S K I N G  ~  version "</span> + VERSION
log.info <span style="color: #98be65;">"reference input    : </span><span style="color: #dcaeea;">${params.reference}</span><span style="color: #98be65;">"</span>
log.info <span style="color: #98be65;">"output directory   : </span><span style="color: #dcaeea;">${params.outdir}</span><span style="color: #98be65;">"</span>
log.info <span style="color: #98be65;">""</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orgafd5440" class="outline-3">
<h3 id="orgafd5440"><span class="section-number-3">2.3</span> Collection of relatively recent LTR retrotranposons</h3>
<div class="outline-text-3" id="text-2-3">
<p>
The rationale for initial collection of recent LTR elements is
that these elements contain less nested insertions and mutations
as well as recombination, thus the chance for false positives is
relatively low. We use the ltrharvest command from <a href="http://genometools.org">genometools</a> to
collect potential LTR sequences with the following
characteristics:
</p>

<ul class="org-ul">
<li>Two terminal repeats which are &gt;= 99% similar, ranging from 100 bp tp 6000 bp</li>
<li>The two terminal repeats end with “TG…CA”</li>
<li>The size of entire element ranges from 1.5 kb to 25 kb</li>
<li>The element must be flanked by a 5bp TSD (target site duplication)</li>
<li>The TSD is within 10 bp from the end of the element</li>
</ul>

<p>
The results are then passed through <code>ltrdigest</code> (again, from
genometools) to retain only those elements with polypurine tract
(PPT) and primer binding site (PBS).
</p>

<p>
The <code>CRL_Step1</code> script ensures that at least 50% of the PPT or PBS
sequences are located in the internal regions and the distance
between LTR and PPT or PBS are no more than 20 bp.
</p>

<p>
The <code>CRL_Step2</code> script filters again. False positives due to gene
clusters may still linger in the output from <code>CRL_Step1</code>. These
instances will have alignable regions that extend past the LTR
boundary. We pull out the flanking sequence from the putative LTR,
and if the flanking sequence <i>also</i> aligns, we discard it.
</p>

<div class="org-src-container">
<pre class="src src-groovy">process recentLTRs {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
          cache <span style="color: #98be65;">'deep'</span>

  input:
  file <span style="color: #98be65;">'genome.fasta'</span> from reference
  file <span style="color: #98be65;">'eukaryotic-tRNAs.fasta.gz'</span> from trnanuc

  output:
  set age, <span style="color: #98be65;">'seqfile.result'</span> into ltrHarvestResultsNew
  set age, <span style="color: #98be65;">'seqfile.result'</span> into ltrHarvestResultsForExemplarNew
  set age, <span style="color: #98be65;">'seqfile.outinner'</span> into ltrHarvestInnerNew
  set age, <span style="color: #98be65;">'seqfile.outinner'</span> into outinnerForBlastXNew
  set age, <span style="color: #98be65;">'CRL_Step2_Passed_Elements.fasta'</span>, <span style="color: #98be65;">'Repeat_down*.fasta'</span>, <span style="color: #98be65;">'Repeat_up*.fasta'</span> into recentLTRs

  script:
  age = <span style="color: #98be65;">'new'</span>
  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">gt suffixerator -db genome.fasta -indexname genome.fasta -tis -suf -lcp -des -ssp -dna</span>

<span style="color: #98be65;">gt ltrharvest \</span>
<span style="color: #98be65;"> -index genome.fasta \</span>
<span style="color: #98be65;"> -out seqfile.out \</span>
<span style="color: #98be65;"> -outinner seqfile.outinner \</span>
<span style="color: #98be65;"> -gff3 seqfile.gff \</span>
<span style="color: #98be65;"> -minlenltr 100 \</span>
<span style="color: #98be65;"> -maxlenltr 6000 \</span>
<span style="color: #98be65;"> -mindistltr 1500 \</span>
<span style="color: #98be65;"> -maxdistltr 25000 \</span>
<span style="color: #98be65;"> -mintsd 5 \</span>
<span style="color: #98be65;"> -maxtsd 5 \</span>
<span style="color: #98be65;"> -motif tgca \</span>
<span style="color: #98be65;"> -similar 99 \</span>
<span style="color: #98be65;"> -vic 10 \</span>
<span style="color: #98be65;">&gt; seqfile.result</span>

<span style="color: #98be65;">gt gff3 \</span>
<span style="color: #98be65;"> -sort seqfile.gff \</span>
<span style="color: #98be65;">&gt; seqfile.gff.sort</span>

<span style="color: #98be65;">zcat eukaryotic-tRNAs.fasta.gz &gt; eukaryotic-tRNAs.fasta</span>

<span style="color: #98be65;">gt ltrdigest \</span>
<span style="color: #98be65;"> -trnas eukaryotic-tRNAs.fasta \</span>
<span style="color: #98be65;"> seqfile.gff.sort \</span>
<span style="color: #98be65;"> genome.fasta \</span>
<span style="color: #98be65;">&gt; seqfile.gff.dgt</span>

<span style="color: #98be65;">CRL_Step1.pl \</span>
<span style="color: #98be65;"> --gff seqfile.gff.dgt</span>

<span style="color: #98be65;">CRL_Step2.pl \</span>
<span style="color: #98be65;"> --step1 CRL_Step1_Passed_Elements.txt \</span>
<span style="color: #98be65;"> --repeatfile seqfile.out \</span>
<span style="color: #98be65;"> --resultfile seqfile.result \</span>
<span style="color: #98be65;"> --sequencefile genome.fasta \</span>
<span style="color: #98be65;"> --removed_repeats CRL_Step2_Passed_Elements.fasta</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>
</div>
</div>

<div id="outline-container-org57db473" class="outline-3">
<h3 id="org57db473"><span class="section-number-3">2.4</span> Collection of relatively old LTR retrotransposons</h3>
<div class="outline-text-3" id="text-2-4">
<p>
Collection of relatively old LTRs is enabled by reducing the
similarity between LTRs to 85% (default value of LTRharvest) and
not associated with terminal sequence motif (but the process is
otherwise identical to <code>recentLTRs</code>).
</p>

<div class="org-src-container">
<pre class="src src-groovy">process olderLTRs {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
          cache <span style="color: #98be65;">'deep'</span>

  input:
  file <span style="color: #98be65;">'genome.fasta'</span> from reference
  file <span style="color: #98be65;">'eukaryotic-tRNAs.fasta.gz'</span> from trnanuc

  output:
  set age, <span style="color: #98be65;">'seqfile.result'</span> into ltrHarvestResultsOld
  set age, <span style="color: #98be65;">'seqfile.result'</span> into ltrHarvestResultsForExemplarOld
  set age, <span style="color: #98be65;">'seqfile.outinner'</span> into ltrHarvestInnerOld
  set age, <span style="color: #98be65;">'seqfile.outinner'</span> into outinnerForBlastXOld
  set age, <span style="color: #98be65;">'CRL_Step2_Passed_Elements.fasta'</span>, <span style="color: #98be65;">'Repeat_down*.fasta'</span>, <span style="color: #98be65;">'Repeat_up*.fasta'</span> into olderLTRs

  script:
  age = <span style="color: #98be65;">'old'</span>
  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">gt suffixerator -db genome.fasta -indexname genome.fasta -tis -suf -lcp -des -ssp -dna</span>

<span style="color: #98be65;">gt ltrharvest \</span>
<span style="color: #98be65;"> -index genome.fasta \</span>
<span style="color: #98be65;"> -out seqfile.out \</span>
<span style="color: #98be65;"> -outinner seqfile.outinner \</span>
<span style="color: #98be65;"> -gff3 seqfile.gff \</span>
<span style="color: #98be65;"> -minlenltr 100 \</span>
<span style="color: #98be65;"> -maxlenltr 6000 \</span>
<span style="color: #98be65;"> -mindistltr 1500 \</span>
<span style="color: #98be65;"> -maxdistltr 25000 \</span>
<span style="color: #98be65;"> -mintsd 5 \</span>
<span style="color: #98be65;"> -maxtsd 5 \</span>
<span style="color: #98be65;"> -vic 10 \</span>
<span style="color: #98be65;">&gt; seqfile.result</span>

<span style="color: #98be65;">gt gff3 \</span>
<span style="color: #98be65;"> -sort seqfile.gff \</span>
<span style="color: #98be65;">&gt; seqfile.gff.sort</span>

<span style="color: #98be65;">zcat eukaryotic-tRNAs.fasta.gz &gt; eukaryotic-tRNAs.fasta</span>

<span style="color: #98be65;">gt ltrdigest \</span>
<span style="color: #98be65;"> -trnas eukaryotic-tRNAs.fasta \</span>
<span style="color: #98be65;"> seqfile.gff.sort \</span>
<span style="color: #98be65;"> genome.fasta \</span>
<span style="color: #98be65;">&gt; seqfile.gff.dgt</span>

<span style="color: #98be65;">CRL_Step1.pl \</span>
<span style="color: #98be65;"> --gff seqfile.gff.dgt</span>

<span style="color: #98be65;">CRL_Step2.pl \</span>
<span style="color: #98be65;"> --step1 CRL_Step1_Passed_Elements.txt \</span>
<span style="color: #98be65;"> --repeatfile seqfile.out \</span>
<span style="color: #98be65;"> --resultfile seqfile.result \</span>
<span style="color: #98be65;"> --sequencefile genome.fasta \</span>
<span style="color: #98be65;"> --removed_repeats CRL_Step2_Passed_Elements.fasta</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>
</div>
</div>

<div id="outline-container-orgbe302cb" class="outline-3">
<h3 id="orgbe302cb"><span class="section-number-3">2.5</span> Cleaning LTR results</h3>
<div class="outline-text-3" id="text-2-5">
<p>
LTRs (both new and old) identified above will almost certainly
include false positives that need to be removed. The most common
errors are:
</p>

<ul class="org-ul">
<li>Tandem local repeats (such as centromeric repeats)</li>
<li>Local gene clusters derived from gene duplications</li>
</ul>

<p>
In the case of genuine LTRs, the insertion site will differ
between LTR instances. The result is that alignment between two
instances will not extend past the borders of the terminal repeat
regions. In false positive instances like the examples above, the
alignability of the instances may extend past the terminal
repeats. :TODO: Present dot-plot examples of true and false LTRs.
</p>

<p>
The outupt of this process (<code>CRL_Step3_Passed_Elements.fasta</code>) is
a FASTA file containing element sequences that have passed the
percent identity (60%) and number of identical nucleotides
thresholds.
</p>

<div class="org-src-container">
<pre class="src src-groovy">ltrs = recentLTRs.mix(olderLTRs)
ltrHarvestResults = ltrHarvestResultsOld.mix(ltrHarvestResultsNew)
ltrHarvestInner = ltrHarvestInnerOld.mix(ltrHarvestInnerNew)
outinnerForBlastX = outinnerForBlastXOld.mix(outinnerForBlastXNew)
ltrHarvestResultsForExemplar = ltrHarvestResultsForExemplarOld.mix(ltrHarvestResultsForExemplarNew)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">process CRL_Step3 {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
  tag { age }

  input:
  set age, <span style="color: #98be65;">'CRL_Step2_Passed_Elements.fasta'</span>, <span style="color: #98be65;">'Repeat_down*.fasta'</span>, <span style="color: #98be65;">'Repeat_up*.fasta'</span> from ltrs

  output:
  set age, <span style="color: #98be65;">'CRL_Step3_Passed_Elements.fasta'</span> into step3Passed
  set age, <span style="color: #98be65;">'CRL_Step3_Passed_Elements.fasta'</span> into step3PassedForExemplars

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">CRL_Step3.pl \</span>
<span style="color: #98be65;"> --directory . \</span>
<span style="color: #98be65;"> --step2 CRL_Step2_Passed_Elements.fasta \</span>
<span style="color: #98be65;"> --pidentity 60 \</span>
<span style="color: #98be65;"> --seq_c 25</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>

<p>
Retrotranposons are frequently nested with each other or inserted
by other elements. If left unidentified, it will cause
misclassification and other complications. To detect those
elements, LTR sequences from candidate elements retained after
steps in 2.1.3 are used to mask the putative internal regions. If
LTR sequences are detected in the internal regions, it is
considered as elements nested with other insertions.
</p>

<p>
The internal regions of elements are also used to search against
a transposase database of DNA transposons. If the internal
sequence has significant matches with any DNA transposase, it is
considered as an element containing nested insertions.
</p>

<p>
This process produces <code>lLTR_Only.lib</code>, a FASTA file containing
the sequence of the left (5'end) LTR sequence.
</p>

<div class="org-src-container">
<pre class="src src-groovy">ltrHarvestResults
.combine(step3Passed, by: 0)
.set { nestedInput }

process identifyNestedInsetions {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
  tag { age }
  input:
  file <span style="color: #98be65;">'genome.fasta'</span> from reference
  set age, <span style="color: #98be65;">'seqfile.result'</span>, <span style="color: #98be65;">'CRL_Step3_Passed_Elements.fasta'</span> from nestedInput

  output:
  set age, <span style="color: #98be65;">'repeats_to_mask_LTR.fasta'</span> into repeatsToMaskLTR

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">ltr_library.pl \</span>
<span style="color: #98be65;"> --resultfile seqfile.result \</span>
<span style="color: #98be65;"> --step3 CRL_Step3_Passed_Elements.fasta \</span>
<span style="color: #98be65;"> --sequencefile genome.fasta</span>
<span style="color: #98be65;">cat lLTR_Only.lib \</span>
<span style="color: #98be65;">| awk 'BEGIN {RS = "&gt;" ; FS = "\\n" ; ORS = ""} \$2 {print "&gt;"\$0}' \</span>
<span style="color: #98be65;">&gt; repeats_to_mask_LTR.fasta</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>
</div>
</div>

<div id="outline-container-orgcae4ec4" class="outline-3">
<h3 id="orgcae4ec4"><span class="section-number-3">2.6</span> Identify elements with nested insertions</h3>
<div class="outline-text-3" id="text-2-6">
<p>
Retrotranposons are frequently nested with each other or inserted
by other elements. If left unidentified, it will cause
misclassification and other complications. To detect those
elements, LTR sequences from candidate elements retained after
steps in == are used to mask the putative internal regions. If
LTR sequences are detected in the internal regions, it is
considered as elements nested with other insertions.
</p>

<div class="org-src-container">
<pre class="src src-groovy">process <span style="color: #ECBE7B;">RepeatMasker1</span> {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
  tag { age }

  input:
  set age, <span style="color: #98be65;">'repeats_to_mask_LTR.fasta'</span>, <span style="color: #98be65;">'seqfile.outinner'</span> from repeatsToMaskLTR.combine(ltrHarvestInner, by: 0)

  output:
  set age, <span style="color: #98be65;">'seqfile.outinner.out'</span>, <span style="color: #98be65;">'seqfile.outinner.masked'</span> into repeatMasker1Unclean

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">RepeatMasker \</span>
<span style="color: #98be65;"> -lib repeats_to_mask_LTR.fasta \</span>
<span style="color: #98be65;"> -nolow \</span>
<span style="color: #98be65;"> -no_is \</span>
<span style="color: #98be65;"> -dir . \</span>
<span style="color: #98be65;"> seqfile.outinner</span>

<span style="color: #98be65;">if [ ! -f seqfile.outinner.masked ]; then</span>
<span style="color: #98be65;">  cp seqfile.outinner seqfile.outinner.masked</span>
<span style="color: #98be65;">fi</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">process cleanRM {
  tag { age }

  input:
  set age, <span style="color: #98be65;">'seqfile.outinner.out'</span>, <span style="color: #98be65;">'seqfile.outinner.masked'</span> from repeatMasker1Unclean

  output:
  set age, <span style="color: #98be65;">'seqfile.outinner.clean'</span> into repeatMasker1Clean

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">cleanRM.pl seqfile.outinner.out seqfile.outinner.masked &gt; seqfile.outinner.unmasked</span>
<span style="color: #98be65;">rmshortinner.pl seqfile.outinner.unmasked 50 &gt; seqfile.outinner.clean</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">process blastX {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
  tag { age }
  cpus 4

  input:
  file <span style="color: #98be65;">'Tpases020812DNA.fasta.gz'</span> from trnaprot
  set age, <span style="color: #98be65;">'seqfile.outinner.clean'</span>, <span style="color: #98be65;">'seqfile.outinner'</span> from repeatMasker1Clean.combine(outinnerForBlastX, by: 0)

  output:
  set age, <span style="color: #98be65;">'passed_outinner_sequence.fasta'</span> into blastxPassed

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">zcat Tpases020812DNA.fasta.gz &gt; Tpases020812DNA.fasta</span>
<span style="color: #98be65;">makeblastdb -in Tpases020812DNA.fasta -dbtype prot</span>
<span style="color: #98be65;">blastx \</span>
<span style="color: #98be65;"> -query seqfile.outinner.clean \</span>
<span style="color: #98be65;"> -db Tpases020812DNA.fasta \</span>
<span style="color: #98be65;"> -evalue 1e-20 \</span>
<span style="color: #98be65;"> -num_descriptions 10 \</span>
<span style="color: #98be65;"> -num_threads </span><span style="color: #dcaeea;">${task.cpus}</span><span style="color: #98be65;"> \</span>
<span style="color: #98be65;"> -out seqfile.outinner.clean_blastx.out.txt</span>

<span style="color: #98be65;">outinner_blastx_parse.pl \</span>
<span style="color: #98be65;"> --blastx seqfile.outinner.clean_blastx.out.txt \</span>
<span style="color: #98be65;"> --outinner seqfile.outinner</span>

<span style="color: #98be65;">if [ ! -s passed_outinner_sequence.fasta ]; then</span>
<span style="color: #98be65;">  echo -e '&gt;dummy empty sequence\nACTACTAC' &gt; passed_outinner_sequence.fasta</span>
<span style="color: #98be65;">fi</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">blastxPassed
.combine(step3PassedForExemplars, by: 0)
.combine(ltrHarvestResultsForExemplar, by: 0)
.set { forExemplarBuilding }

process buildExemplars {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
  tag { age }
  cpus 4

  input:
  file <span style="color: #98be65;">'genome.fasta'</span> from reference
  set age, <span style="color: #98be65;">'passed_outinner_sequence.fasta'</span>, <span style="color: #98be65;">'CRL_Step3_Passed_Elements.fasta'</span>, <span style="color: #98be65;">'seqfile.result'</span> from forExemplarBuilding

  output:
  set age, <span style="color: #98be65;">'LTR.lib'</span> into exemplars

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">CRL_Step4.pl \</span>
<span style="color: #98be65;"> --step3 CRL_Step3_Passed_Elements.fasta \</span>
<span style="color: #98be65;"> --resultfile seqfile.result \</span>
<span style="color: #98be65;"> --innerfile passed_outinner_sequence.fasta \</span>
<span style="color: #98be65;"> --sequencefile genome.fasta</span>

<span style="color: #98be65;">for lib in lLTRs_Seq_For_BLAST.fasta Inner_Seq_For_BLAST.fasta; do</span>
<span style="color: #98be65;">  makeblastdb -in \$lib -dbtype nucl</span>
<span style="color: #98be65;">  blastn \</span>
<span style="color: #98be65;">   -query \${lib} \</span>
<span style="color: #98be65;">   -db \${lib} \</span>
<span style="color: #98be65;">   -evalue 1e-10 \</span>
<span style="color: #98be65;">   -num_threads </span><span style="color: #dcaeea;">${task.cpus}</span><span style="color: #98be65;"> \</span>
<span style="color: #98be65;">   -num_descriptions 1000 \</span>
<span style="color: #98be65;">   -out \${lib}.out</span>
<span style="color: #98be65;">done</span>

<span style="color: #98be65;">CRL_Step5.pl \</span>
<span style="color: #98be65;"> --LTR_blast lLTRs_Seq_For_BLAST.fasta.out \</span>
<span style="color: #98be65;"> --inner_blast Inner_Seq_For_BLAST.fasta.out \</span>
<span style="color: #98be65;"> --step3 CRL_Step3_Passed_Elements.fasta \</span>
<span style="color: #98be65;"> --final LTR.lib \</span>
<span style="color: #98be65;"> --pcoverage 90 \</span>
<span style="color: #98be65;"> --pidentity 80</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>

<p>
Since the set of older LTR elements contain elements from the
newer LTR set, the exemplar sequences need to be masked by
LTR99.lib and all elements that are significantly masked (cutoff
at 80% identity in 90% of the element length) are excluded.
</p>

<div class="org-src-container">
<pre class="src src-groovy">newLTRs = <span style="color: #ECBE7B;">Channel</span>.create()
oldLTRs = <span style="color: #ECBE7B;">Channel</span>.create()

exemplars
.choice( newLTRs, oldLTRs ) { <span style="color: #dcaeea;">it</span>[0] == <span style="color: #98be65;">"new"</span> ? 0 : 1 }

process removeDuplicates {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>

  input:
  set _, <span style="color: #98be65;">'ltrs.new.fasta'</span> from newLTRs
  set _, <span style="color: #98be65;">'ltrs.old.fasta'</span> from oldLTRs

  output:
  set <span style="color: #98be65;">'ltrs.old.fasta.masked'</span>, <span style="color: #98be65;">'ltrs.new.fasta'</span> into bothLTRforMasking

  <span style="color: #98be65;">"RepeatMasker -lib ltrs.new.fasta -dir . ltrs.old.fasta"</span>
}

process filterOldLTRs {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>

  input:
  set <span style="color: #98be65;">'ltrs.old.fasta.masked'</span>, <span style="color: #98be65;">'ltrs.new.fasta'</span> from bothLTRforMasking

  output:
  file <span style="color: #98be65;">'allLTRs.fasta'</span> into allLTR

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">remove_masked_sequence.pl \</span>
<span style="color: #98be65;">--masked_elements ltrs.old.fasta.masked \</span>
<span style="color: #98be65;">--outfile ltrs.old.final.fasta</span>
<span style="color: #98be65;">cat ltrs.new.fasta ltrs.old.final.fasta &gt; allLTRs.fasta</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">allLTR
.splitFasta(record: [id: <span style="color: #a9a1e1;">true</span>, sequence: <span style="color: #a9a1e1;">true</span> ])
.collectFile( name: <span style="color: #98be65;">'allLTRs.fasta'</span> ) { <span style="color: #98be65;">"&gt;"</span> + <span style="color: #dcaeea;">it</span>.id + <span style="color: #98be65;">"#LTR\n"</span> + <span style="color: #dcaeea;">it</span>.sequence }
.tap { allLTR2 }
.set { inputForRM2 }

process <span style="color: #ECBE7B;">RepeatMasker2</span> {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
  cpus 10

  input:
  file <span style="color: #98be65;">'genome.fasta'</span> from reference
  file <span style="color: #98be65;">'allLTR.lib'</span> from inputForRM2

  output:
  file <span style="color: #98be65;">'genome.fasta.masked'</span> into genomeLtrMasked

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">RepeatMasker \</span>
<span style="color: #98be65;"> -no_is \</span>
<span style="color: #98be65;"> -nolow \</span>
<span style="color: #98be65;"> -pa </span><span style="color: #dcaeea;">${task.cpus}</span><span style="color: #98be65;"> \</span>
<span style="color: #98be65;"> -lib allLTR.lib \</span>
<span style="color: #98be65;"> -dir . \</span>
<span style="color: #98be65;"> genome.fasta</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">process <span style="color: #ECBE7B;">RepeatModeler</span> {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
  cpus 4

  input:
  file <span style="color: #98be65;">'genome.masked'</span> from genomeLtrMasked

  output:
  file <span style="color: #98be65;">'consensi.fa.classified'</span> into rmOutput

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">rmaskedpart.pl genome.masked 50 | sed '/^\$/d' &gt; umseqfile</span>
<span style="color: #98be65;">BuildDatabase -name umseqfiledb -engine ncbi umseqfile</span>
<span style="color: #98be65;">RepeatModeler -pa </span><span style="color: #dcaeea;">${task.cpus}</span><span style="color: #98be65;"> -database umseqfiledb &gt;&amp; umseqfile.out</span>
<span style="color: #98be65;">mv RM*/consensi.fa.classified consensi.fa.classified</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">identityUnknown = <span style="color: #ECBE7B;">Channel</span>.create()
identityKnown = <span style="color: #ECBE7B;">Channel</span>.create()

rmOutput
.splitFasta(record: [id: <span style="color: #a9a1e1;">true</span>, text: <span style="color: #a9a1e1;">true</span>])
.choice(identityUnknown, identityKnown) { record -&gt; record.id =~ <span style="color: #98be65;">/#Unknown/</span> ? 0 : 1 }

identityUnknown
.collectFile() { record -&gt; [<span style="color: #98be65;">'unknown.fasta'</span>, record.text] }
.set { repeatmaskerUnknowns }

identityKnown
.collectFile() { record -&gt; [<span style="color: #98be65;">'known.fasta'</span>, record.text] }
.tap { repeatmaskerKnowns1 }
.set{ repeatmaskerKnowns }
</pre>
</div>

<p>
We'd like to get a better idea about what these 'unknown elements'
are. It's possible that they are known transposons that have been
RIPped beyond recognition. To retrieve these, we can pull out an
ailgnment of these repeat instances (with RepeatMasker), run <a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-11-655">DeRIP</a>
over each alignment and then push them back into the identification
pipeline (RepeatModeler classify script).
</p>

<div class="org-src-container">
<pre class="src src-groovy">    process searchForUnidentifiedElements {
      container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>

      input:
      file <span style="color: #98be65;">'genome.fasta'</span> from reference
      file <span style="color: #98be65;">'unknown_elements.fasta'</span> from repeatmaskerKnowns1

      output:
      set <span style="color: #98be65;">'genome.fasta.align'</span>, <span style="color: #98be65;">'unknown_elements.fasta'</span> into unknownAlignments

      <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">    RepeatMasker \</span>
<span style="color: #98be65;">     -lib unknown_elements.fasta \</span>
<span style="color: #98be65;">     -alignments \</span>
<span style="color: #98be65;">     -nolow \</span>
<span style="color: #98be65;">     -no_is \</span>
<span style="color: #98be65;">     -dir . \</span>
<span style="color: #98be65;">     -inv \</span>
<span style="color: #98be65;">     genome.fasta</span>
<span style="color: #98be65;">      """</span>
    }

    process derip {
      container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>

      input:
      set <span style="color: #98be65;">'genome.fasta.align'</span>, <span style="color: #98be65;">'unknown_elements.fasta'</span> from unknownAlignments

      output:
      file <span style="color: #98be65;">'deripped.unknowns.fasta'</span> into derippedUnknowns

      <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">      rmalignment_to_fasta.rb genome.fasta.align unknown_elements.fasta</span>
<span style="color: #98be65;">      for file in alignment*; do</span>
<span style="color: #98be65;">        derip.rb \$file</span>
<span style="color: #98be65;">      done &gt; deripped.unknowns.fasta</span>
<span style="color: #98be65;">      """</span>
    }

    process classifyDeripped {
  container <span style="color: #98be65;">'repeats'</span>

      input:
  file <span style="color: #98be65;">'transposases.fasta.gz'</span> from trnaprot
      file <span style="color: #98be65;">'repeatmodeler_unknowns.deripped.fasta'</span> from derippedUnknowns

  output:
  file <span style="color: #98be65;">'identified_elements.txt'</span> into identifiedDerippedTransposons
      file <span style="color: #98be65;">'unknown_elements.txt'</span> into unidentifiedDerippedTransposons

      <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">zcat transposases.fasta.gz &gt; transposases.fasta</span>
<span style="color: #98be65;">makeblastdb \</span>
<span style="color: #98be65;"> -in transposases.fasta \</span>
<span style="color: #98be65;"> -dbtype prot</span>
<span style="color: #98be65;">blastx \</span>
<span style="color: #98be65;"> -query repeatmodeler_unknowns.deripped.fasta \</span>
<span style="color: #98be65;"> -db transposases.fasta \</span>
<span style="color: #98be65;"> -evalue 1e-10 \</span>
<span style="color: #98be65;"> -num_descriptions 10 \</span>
<span style="color: #98be65;"> -num_threads </span><span style="color: #dcaeea;">${task.cpus}</span><span style="color: #98be65;"> \</span>
<span style="color: #98be65;"> -out modelerunknown_blast_results.txt</span>
<span style="color: #98be65;">transposon_blast_parse.pl \</span>
<span style="color: #98be65;"> --blastx modelerunknown_blast_results.txt \</span>
<span style="color: #98be65;"> --modelerunknown repeatmodeler_unknowns.deripped.fasta</span>
<span style="color: #98be65;">      """</span>
    }

    identifiedDerippedTransposons.subscribe{<span style="color: #51afef;"> println</span>(<span style="color: #98be65;">"Identified, deripped: </span><span style="color: #dcaeea;">${it}</span><span style="color: #98be65;">"</span>) }

</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">process transposonBlast {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
  cpus 4

  input:
  file <span style="color: #98be65;">'transposases.fasta.gz'</span> from trnaprot
  file <span style="color: #98be65;">'repeatmodeler_unknowns.fasta'</span> from repeatmaskerUnknowns

  output:
  file <span style="color: #98be65;">'identified_elements.txt'</span> into identifiedTransposons
      file <span style="color: #98be65;">'unknown_elements.txt'</span> into unidentifiedTransposons

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">zcat transposases.fasta.gz &gt; transposases.fasta</span>
<span style="color: #98be65;">makeblastdb \</span>
<span style="color: #98be65;"> -in transposases.fasta \</span>
<span style="color: #98be65;"> -dbtype prot</span>
<span style="color: #98be65;">blastx \</span>
<span style="color: #98be65;"> -query repeatmodeler_unknowns.fasta \</span>
<span style="color: #98be65;"> -db transposases.fasta \</span>
<span style="color: #98be65;"> -evalue 1e-10 \</span>
<span style="color: #98be65;"> -num_descriptions 10 \</span>
<span style="color: #98be65;"> -num_threads </span><span style="color: #dcaeea;">${task.cpus}</span><span style="color: #98be65;"> \</span>
<span style="color: #98be65;"> -out modelerunknown_blast_results.txt</span>
<span style="color: #98be65;">transposon_blast_parse.pl \</span>
<span style="color: #98be65;"> --blastx modelerunknown_blast_results.txt \</span>
<span style="color: #98be65;"> --modelerunknown repeatmodeler_unknowns.fasta</span>
<span style="color: #98be65;">  """</span>
}

</pre>
</div>
</div>
</div>

<div id="outline-container-orgeadaf53" class="outline-3">
<h3 id="orgeadaf53"><span class="section-number-3">2.7</span> Noncoding RNA</h3>
<div class="outline-text-3" id="text-2-7">
<p>
We'll borrow the excellent ncRNA prediction steps from the
excellent <a href="https://github.com/sanger-pathogens/companion">Companion annotation pipeline</a>.
</p>

<div class="org-src-container">
<pre class="src src-groovy">process predict_ncRNA {
  container <span style="color: #98be65;">'sangerpathogens/companion:latest'</span>

  input:
  file <span style="color: #98be65;">'genome.fasta'</span> from reference

  output:
  file <span style="color: #98be65;">'cm_out'</span> into cmtblouts

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">cp /opt/data/cm/rnas.cm models.cm</span>
<span style="color: #98be65;">cmpress -F models.cm</span>
<span style="color: #98be65;">cmsearch --cpu 1 --tblout cm_out --cut_ga models.cm genome.fasta</span>
<span style="color: #98be65;">  """</span>
}

process merge_ncrnas {
  container <span style="color: #98be65;">'sangerpathogens/companion:latest'</span>
  cache <span style="color: #98be65;">'deep'</span>

  input:
  file <span style="color: #98be65;">'cmtblout'</span> from cmtblouts

  output:
  file <span style="color: #98be65;">'ncrna.gff3'</span> into ncrnafile

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">  infernal_to_gff3.lua &lt; </span><span style="color: #dcaeea;">${cmtblout}</span><span style="color: #98be65;"> &gt; 1</span>
<span style="color: #98be65;">  gt gff3 -sort -tidy -retainids 1 &gt; ncrna.gff3</span>
<span style="color: #98be65;">  """</span>
}

ncrnafile.println()
</pre>
</div>
</div>
</div>

<div id="outline-container-org41d7682" class="outline-3">
<h3 id="org41d7682"><span class="section-number-3">2.8</span> Final Masking</h3>
<div class="outline-text-3" id="text-2-8">
<div class="org-src-container">
<pre class="src src-groovy">repeatmaskerKnowns
.mix(identifiedTransposons)
.collectFile() { <span style="color: #dcaeea;">it</span>.text }
.combine(allLTR2)
.set { knownRepeats }
</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">process repeatMaskerKnowns {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
  publishDir <span style="color: #98be65;">"</span><span style="color: #dcaeea;">${params.outdir}</span><span style="color: #98be65;">/repeatMaskerKnowns"</span>, mode: <span style="color: #98be65;">'copy'</span>

  input:
  file <span style="color: #98be65;">'reference.fasta'</span> from reference
  set <span style="color: #98be65;">'knownTransposons.lib'</span>, <span style="color: #98be65;">'allLTRs.lib'</span> from knownRepeats

  output:
  set <span style="color: #98be65;">'reference.fasta.out'</span>, <span style="color: #98be65;">'reference.fasta.masked'</span> into repeatMaskerKnownsMasked
  set <span style="color: #98be65;">'reference.fasta.out.gff'</span>, <span style="color: #98be65;">'knownRepeats.library.fasta'</span>

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">cat *.lib &gt; knownRepeats.library.fasta</span>
<span style="color: #98be65;">RepeatMasker \</span>
<span style="color: #98be65;"> -lib knownRepeats.library.fasta \</span>
<span style="color: #98be65;"> -nolow \</span>
<span style="color: #98be65;"> -no_is \</span>
<span style="color: #98be65;"> -dir . \</span>
<span style="color: #98be65;"> -gff \</span>
<span style="color: #98be65;"> -s \</span>
<span style="color: #98be65;"> reference.fasta</span>
<span style="color: #98be65;">  """</span>
}

process removeShortMatches {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>
  publishDir <span style="color: #98be65;">"</span><span style="color: #dcaeea;">${params.outdir}</span><span style="color: #98be65;">/cleanMasked"</span>, mode: <span style="color: #98be65;">'copy'</span>

  input:
  file <span style="color: #98be65;">'reference.fa'</span> from reference
  set <span style="color: #98be65;">'rm.out'</span>, <span style="color: #98be65;">'rm.masked'</span> from repeatMaskerKnownsMasked

  output:
  set <span style="color: #98be65;">'rm.trimmed.out'</span>, <span style="color: #98be65;">'rm.trimmed.masked'</span> into repeatMaskerKnownsMaskedTrimmed

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">head -n 3 rm.out &gt; rm.trimmed.out</span>
<span style="color: #98be65;">tail -n +4 rm.out | awk '\$7 - \$6 &gt; </span><span style="color: #dcaeea;">${params.minRepLength}</span><span style="color: #98be65;">' &gt;&gt; rm.trimmed.out</span>
<span style="color: #98be65;">tail -n +4 rm.out | awk 'BEGIN{OFS="\\t"} \$7 - \$6 &gt; </span><span style="color: #dcaeea;">${params.minRepLength}</span><span style="color: #98be65;"> {print \$5, \$6, \$7}' &gt;&gt; rm.trimmed.bed</span>
<span style="color: #98be65;">maskFastaFromBed -fi reference.fa -bed rm.trimmed.bed -fo rm.trimmed.masked -soft</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">process octfta {
  container <span style="color: #98be65;">'robsyme/nf-repeatmasking'</span>

  input:
  file <span style="color: #98be65;">'reference.fa'</span> from reference
  set <span style="color: #98be65;">'rm.out'</span>, <span style="color: #98be65;">'rm.masked'</span> from repeatMaskerKnownsMaskedTrimmed

  output:
  file <span style="color: #98be65;">'summary.tsv'</span> into repeatmaskerSummaryTable

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">build_dictionary.pl --rm rm.out &gt; ltr.dict</span>
<span style="color: #98be65;">one_code_to_find_them_all.pl --rm rm.out --ltr ltr.dict --fasta reference.fa</span>
<span style="color: #98be65;">echo -e 'Family\\tElement Length\\tFragments\\tCopies\\tSolo_LTR\\tTotal_Bp\\tCover\\tchrname' &gt; summary.tsv</span>
<span style="color: #98be65;">for file in *.copynumber.csv; do</span>
<span style="color: #98be65;">  chrname=`echo \$file | sed -e 's/^rm\\.out_//' -e 's/.copynumber.csv\$//'`</span>
<span style="color: #98be65;">  awk -v chrname=\$chrname 'BEGIN{OFS="\\t"} NR&gt;1 &amp;&amp; /^[^#]/ {print(\$0, chrname)}' \$file</span>
<span style="color: #98be65;">done &gt;&gt; summary.tsv</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>
</div>
</div>

<div id="outline-container-org478605c" class="outline-3">
<h3 id="org478605c"><span class="section-number-3">2.9</span> Summary tables and figures</h3>
<div class="outline-text-3" id="text-2-9">
<div class="org-src-container">
<pre class="src src-groovy">process summarise {
  publishDir <span style="color: #98be65;">"</span><span style="color: #dcaeea;">${params.outdir}</span><span style="color: #98be65;">/summarise"</span>, mode: <span style="color: #98be65;">'copy'</span>

  input:
  file <span style="color: #98be65;">'summary.tsv'</span> from repeatmaskerSummaryTable

  output:
  set <span style="color: #98be65;">'summary.bycontig.tidy.tsv'</span>, <span style="color: #98be65;">'summary.tidy.tsv'</span>

  <span style="color: #98be65;">"""</span>
<span style="color: #98be65;">#!/usr/bin/env Rscript</span>
<span style="color: #98be65;">library(ggplot2)</span>
<span style="color: #98be65;">library(dplyr)</span>
<span style="color: #98be65;">library(tidyr)</span>
<span style="color: #98be65;">library(magrittr)</span>

<span style="color: #98be65;">data &lt;- read.table('summary.tsv', header=TRUE) %&gt;%</span>
<span style="color: #98be65;">        separate(Family, into=c("Family", "Subfamily"), sep="/") %&gt;%</span>
<span style="color: #98be65;">        group_by(chrname, Family, Subfamily) %&gt;%</span>
<span style="color: #98be65;">        summarise(fragment.count = sum(Fragments), length = sum(Total_Bp)) %&gt;%</span>
<span style="color: #98be65;">        unite("Family", Family, Subfamily, sep="/")</span>

<span style="color: #98be65;">write.table(data, file='summary.bycontig.tidy.tsv')</span>

<span style="color: #98be65;">data &lt;- read.table('summary.tsv', header=TRUE) %&gt;%</span>
<span style="color: #98be65;">        separate(Family, into=c("Family", "Subfamily"), sep="/") %&gt;%</span>
<span style="color: #98be65;">        group_by(Family, Subfamily) %&gt;%</span>
<span style="color: #98be65;">        summarise(fragment.count = sum(Fragments), length = sum(Total_Bp)) %&gt;%</span>
<span style="color: #98be65;">        unite("Family", Family, Subfamily, sep="/")</span>

<span style="color: #98be65;">write.table(data, file='summary.tidy.tsv', row.names = FALSE)</span>
<span style="color: #98be65;">  """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-groovy">workflow.onComplete {
    log.info <span style="color: #98be65;">"Pipeline completed at: </span><span style="color: #dcaeea;">$workflow.complete</span><span style="color: #98be65;">"</span>
    log.info <span style="color: #98be65;">"Execution status: </span><span style="color: #dcaeea;">${ workflow.success ? 'OK' : 'failed' }</span><span style="color: #98be65;">"</span>
}
</pre>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Robert Syme</p>
<p class="date">Created: 2018-01-03 Wed 16:44</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
